text/3668-async-closure.md

```diff
@@ -0,0 +1,691 @@
+
+For example:
+
+```rust
+let s = String::from("hello, world");
+// Implements `async Fn()` along with `FnMut` and `Fn`
+// because it can copy the `&String` that it captures.
+let _ = async || {
+    println!("{s}");
+};
+
+let s = String::from("hello, world");
+// Implements `async Fn()` but not `FnMut` or `Fn` because
+// it moves and owns a value of type `String`, and therefore
+// the future it returns needs to take a pointer to data
+// owned by the closure.
+let _ = async move || {
+    println!("{s}");
+};
+
+let mut s = String::from("hello, world");
+// Implements `async FnMut()` but not `FnMut` or `Fn`
+// because it needs to reborrow a mutable pointer to `s`.
+let _ = async move || {
+    s.push('!');
+};
+```
+
+#### Specifics about the `AsyncFnOnce` implementation, interaction with `move`
+
+If the closure is inferred to be `async Fn` or `async FnMut`, then the compiler will synthesize an `async FnOnce` implementation for the closure which returns a future that doesn't borrow any captured values from the closure, but instead *moves* the captured values into the future. Synthesizing a distinct future that is returned by `async FnOnce` is necessary because the trait *consumes* the closure when it is called (evident from the `self` receiver type in the method signature), meaning that a self-borrowing future would have references to dropped data. This is an interesting problem described in more detail in [compiler-errors' blog post written on async closures][blog post].
+
+This is reflected in the fact that `AsyncFnOnce::CallOnceFuture` is a distinct type from `AsyncFnMut::CallRefFuture`. While the latter is a generic-associated-type (GAT) due to supporting self-borrows of the called async closure, the former is not, since it must own all of the captures mentioned in the async closures' body.
+
+For example:
+
+```rust
+let s = String::from("hello, world");
+
+let closure = async move || {
+    ready(&s);
+};
+// At this point, `s` is moved out of.  However, the
+// allocation for `s` is still live.  It just lives as a
+// captured field in `closure`.
+
+// Manually call `AsyncFnOnce` -- this isn't stable since
+// `AsyncFnOnce` isn't stable, but it's useful for the demo.
+let fut = AsyncFnOnce::call_once(closure, ());
+// At this point, `closure` is dropped.  However, the
+// allocation for `s` is still live.  It now lives as a
+// captured field in `fut`.
+
+fut.await;
+// After the future is awaited, it's dropped.  At that
+// point, the allocation for `s` is dropped.
+```
+
+Importantly, although these are distinct futures, they still have the same `Output` type (in other words, their futures await to the same type), and for types that have `async Fn*` implementations, the two future types *execute* identically, since they execute the same future body. They only differ in their captures.
```

Comment:
```suggestion
Importantly, although these are distinct futures, they still have the same `Output` type (in other words, their futures await to the same type), and for types that have `async Fn*` implementations, the two future types *execute* identically, since they execute the same future body. They only differ in their captures.

Unlike with the `Fn*` family of traits, there is no way to have a generic implementation of `AsyncFnOnce` for every `AsyncFnMut`.
However, the trait is primarily intended to be implemented by the compiler, so the compiler can just synthesize both implementations.
From a client perspective, it makes sense that if I have an `AsyncFnMut` then I should be able to "call it only once" in a way that is uniform with calling an `AsyncFnOnce`.
Given that clients usually do not care about the concrete future type, but only the final output type, the fact that a different future is used when calling an `AsyncFnMut` via `async_call_mut` vs `async_call_once` will hardly even be noticeable.
```

(I am not sure if this is the best place to put such a comment, feel free to move it around.)

Reply:
I added some clarifying words, but I did not mention how the [ClosureOnceShim](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/enum.InstanceKind.html#variant.ClosureOnceShim) works because that's wholly an implementation detail.

I don't know if anyone knows (and certainly nobody can rely on the fact that) that the actual shim code for the `FnOnce` impl for an `FnMut` closure looks like this:

```
impl FnOnce<A> for {closure} {
  fn call_once(self, args: A) {
    <Self as FnMut<A>>::call_mut(&mut self)
  }
}
```

so it's difficult to contrast that and it feels kinda out of the scope of the RFC.